### ISSUE TITLE
duplicate metrics collector registration attempted

### ISSUE THREAD

**Issue by rtheis** [kubernetes/kubernetes#76956]
**What happened**:

The following error messages were found in the Kubernetes API server and controller manager logs:

```
E0423 17:35:58.491576       1 prometheus.go:138] failed to register depth metric admission_quota_controller: duplicate metrics collector registration attempted
E0423 17:35:58.491625       1 prometheus.go:150] failed to register adds metric admission_quota_controller: duplicate metrics collector registration attempted
E0423 17:35:58.491901       1 prometheus.go:162] failed to register latency metric admission_quota_controller: duplicate metrics collector registration attempted
E0423 17:35:58.492191       1 prometheus.go:174] failed to register work_duration metric admission_quota_controller: duplicate metrics collector registration attempted
E0423 17:35:58.492367       1 prometheus.go:189] failed to register unfinished_work_seconds metric admission_quota_controller: duplicate metrics collector registration attempted
E0423 17:35:58.492507       1 prometheus.go:202] failed to register longest_running_processor_microseconds metric admission_quota_controller: duplicate metrics collector registration attempted
```

These appear to be for deprecated metrics (https://github.com/kubernetes/kubernetes/pull/71300) noted in https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG-1.14.md.

**What you expected to happen**:

No ` duplicate metrics collector registration attempted` error messages in the API server or controller manager logs.

**How to reproduce it (as minimally and precisely as possible)**:

Deploy a Kubernetes version 1.14 cluster and review API server or controller manager logs.

**Anything else we need to know?**:

No.

**Environment**:
- Kubernetes version (use `kubectl version`): 1.14.1
- Cloud provider or hardware configuration: IBM Cloud
- OS (e.g: `cat /etc/os-release`): Ubuntu 16.04.6 LTS
- Kernel (e.g. `uname -a`): 4.4.0-145-generic
- Install tools: Custom
- Network plugin and version (if this is a network-related bug): N/A
- Others: N/A

**Comment by rtheis** [comment]
Updated description to note this impacts controller manager as well.

**Comment by s-urbaniak** [comment]
This is, at least for the API server, because admission options are applied twice during bootstrap, once in https://github.com/kubernetes/kubernetes/blob/59a86822cb966ffb77a702321c89bb4e6f1c1c89/cmd/kube-apiserver/app/apiextensions.go#L53 and once in https://github.com/kubernetes/kubernetes/blob/59a86822cb966ffb77a702321c89bb4e6f1c1c89/cmd/kube-apiserver/app/server.go#L487

While https://github.com/kubernetes/kubernetes/pull/75737 fortunately added `klog.Errorf` for non-deprecated metrics, but we actually have another problem: If admission is reinitialized, the previously (now unused) metric references are still registered, causing the re-registered metrics to be defunct. I didn't check, but essentially this causes workqueue metrics to be defunct right now.

**Comment by logicalhan** [comment]
I think everything is potentially registered up to __*three*__ times (#75779). 

I also do not believe that the metrics are defunct, since I have seen both them and the registration errors co-existing. I could be mistaken but I believe that the way prometheus identifies the metric in the registry (i.e. a hashed value derived from `prometheus.Desc`) means that even if we create two metric objects (i.e. two counters with the same descriptor), that they will be considered identical to the registry. @brancz to verify though, he knows prometheus better than I.

**Comment by s-urbaniak** [comment]
@logicalhan unfortunately the prometheus registry will drop the redundantly registered metrics and return an error. That error is currently just logged.

One way is to specifically analyze the returned error from the registry and use the existing collector reference which I did in #77553. I will also try out using vector metrics instead as it should make the code a bit cleaner.

**Comment by logicalhan** [comment]
What do you mean by drop? Prometheus will not re-register the same metric over again but that is by design.

**Comment by s-urbaniak** [comment]
@logicalhan yes, that is what I meant :-)

**Comment by liminghua999** [comment]
I have the same error when I install k8s 1.14.0 1.14.3   
Do you have the solution for this error?????
`environment：
etcd cluster ： 3 nodes
kernel： 5.1 
os: centos7.6.1810 `
then start 3 node kube-apiserver by ansible
error:
`Jun 24 23:13:25 m-etc-1 systemd[1]: Starting Kubernetes API Server...
Jun 24 23:13:25 m-etc-1 kube-apiserver[13447]: E0624 23:13:25.621114   13447 prometheus.go:138] failed to register depth metric admission_quota_controller: duplicate metrics collector registration attempted
Jun 24 23:13:25 m-etc-1 kube-apiserver[13447]: E0624 23:13:25.621309   13447 prometheus.go:150] failed to register adds metric admission_quota_controller: duplicate metrics collector registration attempted
Jun 24 23:13:25 m-etc-1 kube-apiserver[13447]: E0624 23:13:25.621334   13447 prometheus.go:162] failed to register latency metric admission_quota_controller: duplicate metrics collector registration attempted
Jun 24 23:13:25 m-etc-1 kube-apiserver[13447]: E0624 23:13:25.621410   13447 prometheus.go:174] failed to register work_duration metric admission_quota_controller: duplicate metrics collector registration attempted
Jun 24 23:13:25 m-etc-1 kube-apiserver[13447]: E0624 23:13:25.621433   13447 prometheus.go:189] failed to register unfinished_work_seconds metric admission_quota_controller: duplicate metrics collector registration attempted
Jun 24 23:13:25 m-etc-1 kube-apiserver[13447]: E0624 23:13:25.621450   13447 prometheus.go:202] failed to register longest_running_processor_microseconds metric admission_quota_controller: duplicate metrics collector registration attempted
`

**Comment by zhoujun165** [comment]
I have the same error.
waiting for solution

**Comment by liminghua999** [comment]
> I have the same error.
> waiting for solution

what's  your k8s version??

**Comment by zhoujun165** [comment]
> > I have the same error.
> > waiting for solution
> 
> what's your k8s version??

1.14.3 and 1.15.0

**Comment by liminghua999** [comment]
> > > I have the same error.
> > > waiting for solution
> > 
> > 
> > what's your k8s version??
> 
> 1.14.3 and 1.15.0

I think some wrong with config file ,but I do'nt know where

**Comment by pratsub** [comment]
Same issue I face in centos 7.6 version. kubeadm fails to start. I too eagerly waiting for solution.
k8s version 1.15.1

**Comment by fancygo** [comment]
Waiting for solution!

**Comment by xusxlinux** [comment]
me too
k8s-master-01 kube-apiserver[5127]: E0829 00:37:12.000445    5127 prometheus.go:174] failed to register work_duration metric admission_quota_controller: duplicate metrics collec...ion attempted
k8s-master-01 kube-apiserver[5127]: E0829 00:37:12.000460    5127 prometheus.go:189] failed to register unfinished_work_seconds metric admission_quota_controller: duplicate metr...ion attempted
k8s-master-01 kube-apiserver[5127]: E0829 00:37:12.000469    5127 prometheus.go:202] failed to register longest_running_processor_microseconds metric admission_quota_controller:...ion attempted
k8s-master-01 kube-apiserver[5127]: E0829 00:37:13.530679    5127 prometheus.go:138] failed to register depth metric admission_quota_controller: duplicate metrics collector regi...ion attempted
k8s-master-01 kube-apiserver[5127]: E0829 00:37:13.530703    5127 prometheus.go:150] failed to register adds metric admission_quota_controller: duplicate metrics collector regis...ion attempted
k8s-master-01 kube-apiserver[5127]: E0829 00:37:13.530728    5127 prometheus.go:162] failed to register latency metric admission_quota_controller: duplicate metrics collector re...ion attempted
k8s-master-01 kube-apiserver[5127]: E0829 00:37:13.530746    5127 prometheus.go:174] failed to register work_duration metric admission_quota_controller: duplicate metrics collec...ion attempted
k8s-master-01 kube-apiserver[5127]: E0829 00:37:13.530759    5127 prometheus.go:189] failed to register unfinished_work_seconds metric admission_quota_controller: duplicate metr...ion attempted
k8s-master-01 kube-apiserver[5127]: E0829 00:37:13.530768    5127 prometheus.go:202] failed to register longest_running_processor_microseconds metric admission_quota_controller:...ion attempted
k8s-master-01 kube-apiserver[5127]: E0829 00:37:15.080244    5127 controller.go:148] Unable to remove old endpoints from kubernetes service: StorageError: key not found, Code: 1...onalErrorMsg:

**Comment by Senlian** [comment]
I have the same problem with version  v1.15.2, the log as follows:
```
9月 03 13:16:17 master kube-apiserver[2875]: E0903 13:16:17.741756    2875 prometheus.go:112] failed to register unfinished metric admission_quota_controller: duplicate metrics collector registration attempted
9月 03 13:16:17 master kube-apiserver[2875]: E0903 13:16:17.741765    2875 prometheus.go:126] failed to register unfinished metric admission_quota_controller: duplicate metrics collector registration attempted
9月 03 13:16:17 master kube-apiserver[2875]: E0903 13:16:17.741774    2875 prometheus.go:152] failed to register depth metric admission_quota_controller: duplicate metrics collector registration attempted
9月 03 13:16:17 master kube-apiserver[2875]: E0903 13:16:17.741783    2875 prometheus.go:164] failed to register adds metric admission_quota_controller: duplicate metrics collector registration attempted
9月 03 13:16:17 master kube-apiserver[2875]: E0903 13:16:17.741832    2875 prometheus.go:176] failed to register latency metric admission_quota_controller: duplicate metrics collector registration attempted
9月 03 13:16:17 master kube-apiserver[2875]: E0903 13:16:17.741878    2875 prometheus.go:188] failed to register work_duration metric admission_quota_controller: duplicate metrics collector registration attempted
9月 03 13:16:17 master kube-apiserver[2875]: E0903 13:16:17.741892    2875 prometheus.go:203] failed to register unfinished_work_seconds metric admission_quota_controller: duplicate metrics collector registration attempted
9月 03 13:16:17 master kube-apiserver[2875]: E0903 13:16:17.741902    2875 prometheus.go:216] failed to register longest_running_processor_microseconds metric admission_quota_controller: duplicate metrics collector registration attempted
9月 03 13:16:19 master systemd[1]: Started k8s apiserver.
9月 03 13:16:19 master kube-apiserver[2875]: E0903 13:16:19.410432    2875 controller.go:148] Unable to remove old endpoints from kubernetes service: StorageError: key not found, Code: 1, Key: /registry/masterleases/192.168.159.3, ResourceVersion: 0, AdditionalErrorMsg:
```

**Comment by logicalhan** [comment]
These should be fixed in 1.16.

**Comment by zhangguanzhang** [comment]
`1.14.8` also had this error

**Comment by zhangguanzhang** [comment]
it fix in `v1.16`+

**Comment by ShuLian1984** [comment]
I have the same error.
waiting for solution

system info：
`CentOS Linux release 7.8.2003 (Core)
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7" `

kubernetes version：
`
[root@kuber-22 conf]# kube-apiserver --version
Kubernetes v1.18.2
`
and v1.18.3 too
